beeline.go

```diff
@@ -97,65 +102,146 @@ func Init(config Config) {
+		span.AddField(namespacedKey, val)
+	}
 }
 
-// ContextEvent retrieves the Honeycomb event from a context. You can add fields
-// to the event or override settings (eg sample rate) but should not Send() the
-// event; the wrapper that inserted the event into the Context is responsible
-// for sending it to Hnoeycomb
-func ContextEvent(ctx context.Context) *libhoney.Event {
-	if ctx != nil {
-		if evt, ok := ctx.Value(honeyEventContextKey).(*libhoney.Event); ok {
-			return evt
-		}
+// AddRollupFieldToSpan allows you to add a numeric field to the current span,
+// and, when called on multiple spans within a trace, the sum of the field will
+// be added to the root span. Use this when doing an action many times or on
+// many spans and you want the sum of all those actions to be represented on the
+// root span. Fields added here are prefixed with `app.` and the rolled up
+// fields on the root span are prefixed with `totals.app.`
+func AddRollupFieldToSpan(ctx context.Context, key string, val float64) {
+	namespacedKey := fmt.Sprintf("app.%s", key)
+	span := internal.CurrentSpan(ctx)
+	if span != nil {
+		span.AddRollupField(namespacedKey, val)
 	}
-	return nil
 }
 
-// contextBuilder isn't used yet but matches ContextEvent. When it's useful,
-// export it, but until then it's just confusing.
-func contextBuilder(ctx context.Context) *libhoney.Builder {
-	if bldr, ok := ctx.Value(honeyBuilderContextKey).(*libhoney.Builder); ok {
-		return bldr
+// AddFieldToTrace adds the field to both the currently active span and all
+// other spans involved in this trace that occur within this process.
+// Additionally, these fields are packaged up and passed along to downstream
+// processes if they are also using a beeline. This function is good for adding
+// context that is better scoped to the request than this specific unit of work,
+// eg user IDs, globally relevant feature flags, errors, etc. Fields added here
+// are prefixed with `global.`
+func AddFieldToTrace(ctx context.Context, key string, val interface{}) {
+	namespacedKey := fmt.Sprintf("global.%s", key)
+	trace := internal.GetTraceFromContext(ctx)
+	if trace != nil {
+		trace.AddField(namespacedKey, val)
 	}
-	return nil
+}
+
+// HasTrace returns true if there is a trace in the current context
+func HasTrace(ctx context.Context) bool {
+	trace := internal.GetTraceFromContext(ctx)
+	return trace != nil
+}
+
+// StartTraceWithIDs lets you start a trace with a specific set of IDs - it is
+// used when you've received the IDs from another source (eg incoming HTTP
+// headers). If you don't care what the IDs are, you may use either this or
+// StartSpan to start a trace. You cannot change the trace IDs after a trace has
+// begun. StartTraceWithIDs also starts a span (the root span) for this trace.
+// You should call FinishSpan to close the span (and trace) started by this
+// function.
+func StartTraceWithIDs(ctx context.Context, traceID, parentID, name string) context.Context {
```

Comment:
What do you think about allowing this to take a map of fields as well? Given our trace context format includes fields to be propagated, it'd be nice if our function for starting a new trace supported those too, instead of having to do a separate `AddFieldToTrace` loop.

Unrelatedly, what happens if you call `StartTraceWithIDs` when there is already an active trace? I can't find the line reference now, but I think the node beeline reports an error in this case.